contracts/vesting.token/src/vesting.token.cpp

```diff
@@ -0,0 +1,101 @@
+        vesting_category category = category_iter->second;
+
+
+        // Check the symbol is correct and valid
+        auto sym = amount.symbol;
+        eosio::check(sym == system_resource_currency, "Symbol does not match system resource currency");
+        eosio::check(sym.precision() == system_resource_currency.precision(), "Symbol precision does not match");
+        eosio::check(amount.amount > 0, "Amount must be greater than 0");
+
+        // Create a new vesting schedule
+        vesting_allocations vesting_table(get_self(), holder.value);
+        auto iter = vesting_table.end();
+
+        // Check the number of rows for the user
+        uint32_t num_rows = 0;
+        for (auto itr = vesting_table.begin(); itr != vesting_table.end(); ++itr) {
+            ++num_rows;
+            if (num_rows > 20) {
+                eosio::check(false, "Cannot purchase tokens more than 20 times.");
+            }
+        }
+
+        // Calculate the number of seconds since sales start
+        const eosio::time_point_sec now = eosio::current_time_point();
+        int64_t allocated_after_sales_start_seconds = now.sec_since_epoch() - sales_start_date.sec_since_epoch();
+
+        vesting_table.emplace(get_self(), [&](auto& row) {
+            row.holder = holder;
+            row.total_allocated = amount;
+            row.tokens_claimed = asset(0, amount.symbol);
+            row.allocated = allocated_after_sales_start_seconds;
+            row.vesting_category_type = category;
+        });
+    }
+    void vestingToken::withdraw(eosio::name holder) {
+        require_auth(holder);
+
+        // Get the vesting allocations
+        vesting_allocations vesting_table(get_self(), holder.value);
+        const eosio::time_point_sec now = eosio::current_time_point();
+
+        for (auto iter = vesting_table.begin(); iter != vesting_table.end(); ++iter) {
+            const vested_allocation& vesting_allocation = *iter;
+
+            vesting_category category = vesting_allocation.vesting_category_type;
+
+            uint32_t vesting_start_since_epoch = launch_date.sec_since_epoch() + vesting_allocation.allocated + (category.start_delay_days * SECONDS_IN_DAY);
+            uint32_t cliff_end_since_epoch = vesting_start_since_epoch + (category.cliff_period_days * SECONDS_IN_DAY);
+            
+            // Calculate the vesting end time
+            uint32_t vesting_end_time_since_epoch = vesting_start_since_epoch + (category.vesting_period_days * SECONDS_IN_DAY);
+
+            // Check if vesting period after cliff has started
+            eosio::check(now.sec_since_epoch() >= cliff_end_since_epoch, "Vesting period after cliff has not started");
+
+            // Check if vesting period after cliff has started
+            eosio::check(now.sec_since_epoch() >= vesting_end_time_since_epoch, "Vesting period after cliff has not started");
+
+            eosio::asset cliff_tokens = vesting_allocation.total_allocated * category.cliff_period_days /
```

Comment:
vesting_allocation.total_allocated is an https://docs.antelope.io/cdt/latest/reference/Classes/structeosio_1_1asset asset class, not a number. I don't think it has arithmetic overloads so don't know if this formula works with type differences. It could work, its worth a try I guess.

You might want to use the asset::amount member variable to carry out the arithmetic instead.

Reply:
this also applies to L82, L85, 89, 95

Reply:
/contracts/src/vesting.token.cpp:84:27: error: invalid use of non-static data member 'amount'
            eosio::asset::amount cliff_tokens = vesting_allocation.total_allocated * category.cliff_period_seconds /
            ~~~~~~~~~~~~~~^~~~~~
/contracts/src/vesting.token.cpp:88:27: error: invalid use of non-static data member 'amount'
            eosio::asset::amount claimable = (vesting_allocation.total_allocated - cliff_tokens) * seconds_elapsed / category.vesting_period_seconds;
            ~~~~~~~~~~~~~~^~~~~~